### PR TITLE
[HWArithToHW] Lower Cast,Constant,Add,Sub,Mul,Div

### PR DESCRIPTION
**Pull request by 7FM**
As promised in https://github.com/llvm/circt/pull/3546, here is the follow up PR that provides the lowering patterns as well as tests.

CC @jopperm @mortbopet

### CODE REVIEW

File: lib/Conversion/HWArithToHW/HWArithToHW.cpp [hunk 1]
```diff
@@ -15,26 +15,212 @@
...
+namespace {
+struct DivOpLowering : public OpConversionPattern<DivOp> {
+  using OpConversionPattern<DivOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(DivOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto lhsType = op.inputs()[0].getType().template cast<IntegerType>();
+    auto rhsType = op.inputs()[1].getType().template cast<IntegerType>();
```
Comment (mortbopet) [inline lib/Conversion/HWArithToHW/HWArithToHW.cpp:92]: nit: `op.inputs()[0] -> op.getOperand(0)`.

File: lib/Conversion/HWArithToHW/HWArithToHW.cpp [hunk 2]
```diff
@@ -15,26 +15,212 @@
...
+    auto targetType = op.result().getType().template cast<IntegerType>();
+
+    // comb.div* needs identical bitwidths for its operands and their result.
```
Comment (mortbopet) [inline lib/Conversion/HWArithToHW/HWArithToHW.cpp:95]: wording: **its** operands and **their** result

File: lib/Conversion/HWArithToHW/HWArithToHW.cpp [hunk 3]
```diff
@@ -15,26 +15,212 @@
...
+    // The bitwidth for RHS is bit tricky. If the RHS is unsigned and we are
+    // about to perform a signed division, then we need one additional bit to
+    // avoid misinterpretation of RHS as a signed value!
+    bool signedDivision = targetType.isSigned();
+    unsigned extendSize = std::max(
+        targetType.getWidth(),
+        rhsType.getWidth() + (signedDivision && !rhsType.isSigned() ? 1 : 0));
+
+    // Extend the operands
+    Value lhsValue = extendTypeWidth(rewriter, loc, adaptor.inputs()[0],
```
Comment (mortbopet) [inline lib/Conversion/HWArithToHW/HWArithToHW.cpp:110]: `getOperand(#)`, if possible.

File: lib/Conversion/HWArithToHW/HWArithToHW.cpp [hunk 4]
```diff
@@ -15,26 +15,212 @@
...
+    if (signedDivision)
+      divResult = rewriter.create<comb::DivSOp>(loc, lhsValue, rhsValue)
+                      ->getOpResult(0);
+    else
+      divResult = rewriter.create<comb::DivUOp>(loc, lhsValue, rhsValue)
+                      ->getOpResult(0);
+
+    // finally truncate back to the expected result size!
+    Value truncateResult =
+        extractBits(rewriter, loc, divResult, 0, targetType.getWidth());
```
Comment (mortbopet) [inline lib/Conversion/HWArithToHW/HWArithToHW.cpp:125]: Nit: add `/*...=*/` descriptions to arguments (random `0` requires people to jump back to `extractBits` to check what it implies).

File: lib/Conversion/HWArithToHW/HWArithToHW.cpp [hunk 5]
```diff
@@ -15,26 +15,212 @@
...
+
+namespace {
+struct CastOpLowering : public OpConversionPattern<CastOp> {
+  using OpConversionPattern<CastOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(CastOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto sourceType = op.in().getType().cast<IntegerType>();
+    auto targetType = op.out().getType().cast<IntegerType>();
```
Comment (mortbopet) [inline lib/Conversion/HWArithToHW/HWArithToHW.cpp:141]: suggestion: cache `sourceWidth, targetWidth` instead - these are more used than the types in this function.

File: lib/Conversion/HWArithToHW/HWArithToHW.cpp [hunk 6]
```diff
@@ -15,26 +15,212 @@
...
+template <class BinOp, class ReplaceOp>
+struct BinaryOpLowering : public OpConversionPattern<BinOp> {
+  using OpConversionPattern<BinOp>::OpConversionPattern;
+  using OpAdaptor = typename OpConversionPattern<BinOp>::OpAdaptor;
+
+  LogicalResult
+  matchAndRewrite(BinOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto lhsType = op.inputs()[0].getType().template cast<IntegerType>();
```
Comment (mortbopet) [inline lib/Conversion/HWArithToHW/HWArithToHW.cpp:178]: `getOperand(...)`

**Review by mortbopet — APPROVED** [review]
Code LGTM with a few nits in the code and tests.

File: lib/Conversion/HWArithToHW/HWArithToHW.cpp (hunk 3 above)
Comment (7FM) [reply lib/Conversion/HWArithToHW/HWArithToHW.cpp:110]: Unfortunately, the `OpAdaptor`'s do not seem to have this accessor.

**Pull request merged**